rfcs/0091-pure-object-prototypes.md

```diff
@@ -0,0 +1,418 @@
+};
+```
+
+Another extension might instead be about using the CCL implementation instead of SBCL,
+which includes a different default debug level that is only valid if there are no overrides:
+```
+useCcl = {
+  extension = self: super: {
+    implementation = pkgs.ccl;
+  };
+  defaults = {
+    debugLevel = 1;
+  };
+  supers = [ clApplication ];
+};
+```
+
+Meanwhile, another extension might be about using a graphical debugging,
+including a graphical inspector:
+```
+clGraphicalDebugging = {
+  extension = self: super: {
+    extraSystems = super.extraSystems ++ [ lispSystems.clouseau ];
+  };
+  supers = [ clDebugging ];
+};
+```
+
+In the end you can define your application:
+```
+ernestine = pop {
+  extension = self: super: {
+    system = "ernestine-gui";
+  };
+  supers = [ clApplication ];
+};
+```
+
+And you can define a debugging variant of your application:
+```
+ernestine = pop {
+  supers = [ super.ernestine clGraphicalDebugging ];
+}
+```
+
+The multiple-inheritance mechanism ensures that defaults override other defaults but not extensions;
+it also ensures that extensions are evaluated once and only once, in dependency order,
+so that if a dependency appears many times, if doesn't get to re-do its changes
+and undo those of other dependencies.
+
+Of course, there are library functions that simplify the cases where a POP
+only has an `extension`, only has a `defaults`, or only has `supers`,
+but we avoided using them for the sake of this example.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It's a change:
+  every change is disruptive.
+  If we want to embrace OOP, then the object system we adopt will become pervasive,
+  and clash with the large body of existing code.
+  Maybe then we should implement that autodetection magic so that POP can
+  seamlessly interoperate with previous extension systems
+  in an incremental embrace-and-extend replacement.
+
+- Experimental:
+  It's working well for me, but so far, I'm the only user.
+  Quite possibly, the UI (name and signature of toplevel functions)
+  might be adjusted based on feedback by other users.
+  Also, if people actively embrace OOP, they may want to add more features to POP
+  (multi-methods, method combinations, meta-object protocol, etc.),
+  at which point the design may have to evolve.
+  The library could be put in a staging area until it's considered stable.
+  See [RFC 0082 lib.experimental](https://github.com/NixOS/rfcs/pull/82).
+
+- Performance:
+  Computing the `precedenceList` for an object in general is
+  linear in the size of the inheritance graph (nodes plus arcs),
+  i.e. the number of transitive `supers` entries.
+  In the worst case, that's up to quadratic for each new object,
+  which is cubic in the total number of objects.
+  That can be slow.
+  That said, the same "computation" would have to be done by hand
+  by users who would want to achieve the same effect without automation,
+  so it's not really a drawback *given the desired effect*.
```

Comment:
… but automating handling this might encourage larger dependency graphs, making performance an issue.

(also, manual handling usually needs to be done once a few months with the result of the reevaluation «cached» in the form of an expression — here the cost is per-evaluation)

Reply:
I added these two remarks in the drawback discussion.